### PR TITLE
Predicate regex

### PR DESCRIPTION
**Pull request by martin-volf**

### CODE REVIEW

File: src/confd_gnmi_common.py [hunk 1]
```diff
@@ -150,10 +157,16 @@ def make_path(gnmi_path):
...
-                    val)
+            len_items = len(e.key.items())
+            if len_items:
+                if not xpath: path += '{'
+                for index, (k, v) in enumerate(e.key.items()):
+                    val = v if not quote_val else "\"{}\"".format(v)
+                    path += "[{}={}]".format(k, val) if xpath else "{}".format(
+                        val)
+                    if not xpath and index < len_items - 1: path += " "
+                if not xpath: path += '}'
```
Comment (martin-volf) [inline src/confd_gnmi_common.py:168]: ouch... this unreadable code still here...

**Review by micnovak — APPROVED** [review]

File: gnmi-tools.yang [hunk 2]
```diff
@@ -32,6 +32,13 @@ module gnmi-tools {
       leaf name {type string;}
       uses combo;
     }
+
+    list double-list {
```
Comment (jomiklos) [inline gnmi-tools.yang:36]: imho `double(-)key-list` / `two-keys-list` / etc. would be better, less unclear name?

File: src/confd_gnmi_common.py [hunk 3]
```diff
@@ -136,27 +143,31 @@ def split_gnmi_path(xpath_string: str, prefix_len: int) -> Tuple[str, str]:
...
-                val = v if not quote_val else "\"{}\"".format(v)
-                path += "[{}={}]".format(k, val) if xpath else "{{{}}}".format(
-                    val)
-        if path == "":
-            path = "/"
-        return path
+        path_elements = [e.name + elem_keys(e) for e in gnmi_path.elem]
+        return "/" + "/".join(path_elements)
+
+    valstr = '"{}"' if quote_val else '{}'
```
Comment (jomiklos) [inline src/confd_gnmi_common.py:160]: feels a bit lost in between two function declarations...

File: tests/client_server_test_base.py [hunk 4]
```diff
@@ -84,7 +90,13 @@ def capability_supported(cap):
 
     @staticmethod
     def assert_update(update, path_val):
-        assert (update.path == path_val[0])
+        check_path = True
+        #do we have options attribute?
```
Comment (jomiklos) [inline tests/client_server_test_base.py:94]: do we?

File: tests/test_client_server_confd.py [hunk 5]
```diff
@@ -76,6 +76,11 @@ def _test_gnmi_tools_get_subscribe_gnmi_tools(self, is_subscribe=False,
         kwargs = {"assert_fun": GrpcBase.assert_updates}
         kwargs["prefix"] = make_gnmi_path("/gnmi-tools:gnmi-tools")
 
+        leaf_paths_val = [
+            (GrpcBase.mk_gnmi_if_path(p[0]), p[1]) if len(p) == 2 else (
+                GrpcBase.mk_gnmi_if_path(p[0]), p[1], p[2]) for p in
+            self.gnmi_tools_leaf_paths_str_val]
```
Comment (jomiklos) [inline tests/test_client_server_confd.py:82]: `gnmi_tools_leaf_paths_str_val` sounds a bit too much chatty variable name imho, though i understand this is just its use, not declaration in this PR...

**Pull request merged**